### PR TITLE
added mutexes to prevent race conditions

### PR DESCRIPTION
**Pull request by paulhovey**
Fixes for #164

### CODE REVIEW

File: widget/entry.go [hunk 1]
```diff
@@ -84,13 +84,12 @@ func (e *entryRenderer) BackgroundColor() color.Color {
...
 func (e *entryRenderer) Refresh() {
-	e.placeholder.refreshTextRenderer()
-	e.placeholder.Hide()
 	if e.text.len() == 0 && e.entry.Visible() {
 		e.placeholder.Show()
+	} else if e.placeholder.Visible() {
+		e.placeholder.Hide()
 	}
 
-	e.text.refreshTextRenderer()
```
Comment (andydotxyz) [inline widget/entry.go]: Is there no ramification or taking out this (and the other, above) renderer refresh calls?

Comment (paulhovey) [reply widget/entry.go]: I didn't notice any.  The original debugging showed that `textProvider.Refresh()` was getting called 3 times for every input, and my thinking was if I got that down to 1 that would solve the problem.  It didn't, but seemed like the right approach to making it better regardless.

Comment (andydotxyz) [reply widget/entry.go]: Ok, thanks - just wanted to check

File: widget/text.go [hunk 2]
```diff
@@ -192,6 +200,7 @@ type textRenderer struct {
 	objects []fyne.CanvasObject
 
 	texts []*canvas.Text
+	tLock sync.RWMutex
```
Comment (andydotxyz) [inline widget/text.go]: I don’t know what “tLock” means - is should it just be “lock” or the more specific “textLock”?

**Comment by paulhovey** [comment]
effff I brought in the develop branch, thinking that's what the PR was against.

**Comment by paulhovey** [comment]
After testing with https://github.com/fyne-io/fyne/pull/179/commits/c3b13d66544e66971a598d9d3858116429d4d68b , I was unable to see any race conditions after testing.

**Comment by andydotxyz** [comment]
The settings and base object changes are clean and simple.
However the textProvider changes seem to have bug fixes interspersed and the lock semantics are leaking out to objects that use it... I don't know if I am comfortable merging all of that in one go...

Could we PR for the first two and the bugs in textProivder then continue to look at locking in the provider afterwards?

**Comment by paulhovey** [comment]
I'm closing this because @andydotxyz wanted the bugfixes separate from the mutexes.  @steveoc64 did a bunch of mutexes in #190 , I added the entry mutex in #196 , and the bugfixes in #195 .

**Pull request closed without merging**